[tacmi] Error parsing API Scheme: Non-terminating decimal expansion; no exact representable decimal result

## Error:
I am receiving the following error when parsing my Schema API page, if:
 * the parameter being parsed has a unit of "l/h" from the CMI   -- and --
 * the value is !=0
The value causing the error and all subsequent values are not updated


## Item Definition
The item definition is as follows:
    Number:VolumetricFlowRate D_V_ST "Solarthermie durchfluss" {channel="tacmi:cmiSchema:ratiotherm:dV_ST"}
I also tried:
    Number:Volume D_V_ST "Solarthermie durchfluss" {channel="tacmi:cmiSchema:ratiotherm:dV_ST"}
        --and--
    Number D_V_ST "Solarthermie durchfluss" {channel="tacmi:cmiSchema:ratiotherm:dV_ST"}
without change of behaviour

## Trace
Trace is the following (with one working parameter showing before the actual error):
```
2022-03-08 08:54:23.067 [TRACE] [.tacmi.internal.schema.ApiPageParser] - Found parameter pos25:78:8 [READ_ONLY] : temp_ST_RL "Solarthermie Ruecklauf" = 24.9 °C
2022-03-08 08:54:23.070 [TRACE] [.tacmi.internal.schema.ApiPageParser] - Found parameter pos26:81:8 [READ_ONLY] : dV_ST "Solarthermie Durchfluss" = 466 l/h
2022-03-08 08:54:23.071 [DEBUG] [i.internal.schema.TACmiSchemaHandler] - Error parsing API Scheme: Non-terminating decimal expansion; no exact representable decima
l result.
org.attoparser.ParseException: Non-terminating decimal expansion; no exact representable decimal result.
        at org.attoparser.MarkupParser.parseDocument(MarkupParser.java:393) ~[?:?]
        at org.attoparser.MarkupParser.parse(MarkupParser.java:257) ~[?:?]
        at org.attoparser.MarkupParser.parse(MarkupParser.java:175) ~[?:?]
        at org.attoparser.simple.SimpleMarkupParser.parse(SimpleMarkupParser.java:132) ~[?:?]
        at org.openhab.binding.tacmi.internal.schema.TACmiSchemaHandler.parsePage(TACmiSchemaHandler.java:160) ~[?:?]
        at org.openhab.binding.tacmi.internal.schema.TACmiSchemaHandler.refreshData(TACmiSchemaHandler.java:170) ~[?:?]
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515) [?:?]
        at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:305) [?:?]
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:305) [?:?]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
        at java.lang.Thread.run(Thread.java:829) [?:?]
Caused by: java.lang.ArithmeticException: Non-terminating decimal expansion; no exact representable decimal result.
        at java.math.BigDecimal.divide(BigDecimal.java:1722) ~[?:?]
        at org.openhab.binding.tacmi.internal.schema.ApiPageParser.getApiPageEntry(ApiPageParser.java:388) ~[?:?]
        at org.openhab.binding.tacmi.internal.schema.ApiPageParser.handleCloseElement(ApiPageParser.java:210) ~[?:?]
        at org.attoparser.simple.SimplifierMarkupHandler.handleCloseElementEnd(SimplifierMarkupHandler.java:341) ~[?:?]
        at org.attoparser.MarkupEventProcessorHandler.handleCloseElementEnd(MarkupEventProcessorHandler.java:473) ~[?:?]
        at org.attoparser.ParsingElementMarkupUtil.parseCloseElement(ParsingElementMarkupUtil.java:201) ~[?:?]
        at org.attoparser.MarkupParser.parseBuffer(MarkupParser.java:725) ~[?:?]
        at org.attoparser.MarkupParser.parseDocument(MarkupParser.java:301) ~[?:?]
        ... 11 more
```
## Environment
- Openhab 3.2 release with corresponding TACMI binding on Debian Buster
- openjdk 11.0.13 2021-10-19 LTS

Thanks a lot for your support here
@marvkis, would be amazing if you could take a look, happy to help with further debugging or testing in my setup.

thanks and best regards,
/Michael

--- a/bundles/org.openhab.binding.tacmi/src/main/java/org/openhab/binding/tacmi/internal/schema/ApiPageParser.java
+++ b/bundles/org.openhab.binding.tacmi/src/main/java/org/openhab/binding/tacmi/internal/schema/ApiPageParser.java
@@ -15,6 +15,10 @@ package org.openhab.binding.tacmi.internal.schema;
 import java.math.BigDecimal;
 import java.net.URI;
 import java.net.URISyntaxException;
+import java.time.LocalDate;
+import java.time.LocalTime;
+import java.time.ZoneId;
+import java.time.format.DateTimeFormatter;
 import java.util.ArrayList;
 import java.util.HashSet;
 import java.util.List;
@@ -33,6 +37,7 @@ import org.eclipse.jetty.util.StringUtil;
 import org.openhab.binding.tacmi.internal.TACmiBindingConstants;
 import org.openhab.binding.tacmi.internal.TACmiChannelTypeProvider;
 import org.openhab.binding.tacmi.internal.schema.ApiPageEntry.Type;
+import org.openhab.core.library.types.DateTimeType;
 import org.openhab.core.library.types.DecimalType;
 import org.openhab.core.library.types.OnOffType;
 import org.openhab.core.library.types.QuantityType;
@@ -196,7 +201,18 @@ public class ApiPageParser extends AbstractSimpleMarkupHandler {
                     sb = sb.delete(0, 0);
                 }
                 if (this.fieldType == FieldType.READ_ONLY || this.fieldType == FieldType.FORM_VALUE) {
+                    int len = sb.length();
                     int lids = sb.lastIndexOf(":");
+                    if (len - lids == 3) {
+                        int lids2 = sb.lastIndexOf(":", lids - 1);
+                        if (lids2 > 0 && (lids - lids2 >= 3 && lids - lids2 <= 7)) {
+                            // the given value might be a time. validate it
+                            String timeCandidate = sb.substring(lids2 + 1).trim();
+                            if (timeCandidate.length() == 5 && timeCandidate.matches("[0-9]{2}:[0-9]{2}")) {
+                                lids = lids2;
+                            }
+                        }
+                    }
                     int fsp = sb.indexOf(" ");
                     if (fsp < 0 || lids < 0 || fsp > lids) {
                         logger.debug("Invalid format for setting {}:{}:{} [{}] : {}", id, line, col, this.fieldType,
@@ -401,16 +417,27 @@ public class ApiPageParser extends AbstractSimpleMarkupHandler {
                             type = Type.NUMERIC_FORM;
                         }
                     } catch (NumberFormatException nfe) {
-                        // not a number...
-                        channelType = "String";
+                        ctuid = null;
+                        // check for time....
+                        String[] valParts = vs.split(":");
+                        if (valParts.length == 2) {
+                            channelType = "DateTime";
+                            // convert it to zonedDateTime with today as date and the
+                            // default timezone.
+                            var zdt = LocalTime.parse(vs, DateTimeFormatter.ofPattern("HH:mm")).atDate(LocalDate.now())
+                                    .atZone(ZoneId.systemDefault());
+                            state = new DateTimeType(zdt);
+                            type = Type.NUMERIC_FORM;
+                        } else {
+                            // not a number and not time...
+                            channelType = "String";
+                            state = new StringType(vs);
+                            type = Type.STATE_FORM;
+                        }
                         if (this.fieldType == FieldType.READ_ONLY || this.address == null) {
                             ctuid = TACmiBindingConstants.CHANNEL_TYPE_SCHEME_STATE_RO_UID;
                             type = Type.READ_ONLY_STATE;
-                        } else {
-                            ctuid = null;
-                            type = Type.STATE_FORM;
                         }
-                        state = new StringType(vs);
                     }
                 }
                 break;
@@ -555,8 +582,11 @@ public class ApiPageParser extends AbstractSimpleMarkupHandler {
                     }
                 }
                 break;
+            case TIME:
+                itemType = "DateTime";
+                break;
             default:
-                throw new IllegalStateException();
+                throw new IllegalStateException("Unhandled OptionType: " + cx2e.optionType);
         }
         ChannelTypeBuilder<?> ctb = ChannelTypeBuilder
                 .state(new ChannelTypeUID(TACmiBindingConstants.BINDING_ID, shortName), shortName, itemType)

--- a/bundles/org.openhab.binding.tacmi/src/main/java/org/openhab/binding/tacmi/internal/schema/ApiPageParser.java
+++ b/bundles/org.openhab.binding.tacmi/src/main/java/org/openhab/binding/tacmi/internal/schema/ApiPageParser.java
@@ -350,7 +350,7 @@ public class ApiPageParser extends AbstractSimpleMarkupHandler {
                         // for the older pre-X2 devices (i.e. the UVR 1611) we get a comma. So we
                         // we replace all ',' with '.' to check if it's a valid number...
                         String val = valParts[0].replace(',', '.');
-                        BigDecimal bd = new BigDecimal(val);
+                        float bd = Float.parseFloat(val);
                         if (valParts.length == 2) {
                             if ("°C".equals(valParts[1])) {
                                 channelType = "Number:Temperature";
@@ -374,15 +374,14 @@ public class ApiPageParser extends AbstractSimpleMarkupHandler {
                                 state = new QuantityType<>(bd, Units.HERTZ);
                             } else if ("kW".equals(valParts[1])) {
                                 channelType = "Number:Power";
-                                bd = bd.multiply(new BigDecimal(1000));
+                                bd = bd *= 1000;
                                 state = new QuantityType<>(bd, Units.WATT);
                             } else if ("kWh".equals(valParts[1])) {
-                                channelType = "Number:Power";
-                                bd = bd.multiply(new BigDecimal(1000));
+                                channelType = "Number:Energy";
                                 state = new QuantityType<>(bd, Units.KILOWATT_HOUR);
                             } else if ("l/h".equals(valParts[1])) {
-                                channelType = "Number:Volume";
-                                bd = bd.divide(new BigDecimal(60));
+                                channelType = "Number:VolumetricFlowRate";
+                                bd = bd /= 60;
                                 state = new QuantityType<>(bd, Units.LITRE_PER_MINUTE);
                             } else {
                                 channelType = "Number";
@@ -440,7 +439,29 @@ public class ApiPageParser extends AbstractSimpleMarkupHandler {
                     logger.warn("Error loading API Scheme: {} ", ex.getMessage());
                 }
             }
-            if (channel == null || !Objects.equals(ctuid, channel.getChannelTypeUID())) {
+            if (e != null && !channelType.equals(e.channel.getAcceptedItemType())) {
+                // channel type has changed. we have to rebuild the channel.
+                this.channels.remove(channel);
+                channel = null;
+            }
+            if (channel != null && ctuid == null && cx2e != null) {
+                // custom channel type - check if it already exists and recreate when needed...
+                ChannelTypeUID curCtuid = channel.getChannelTypeUID();
+                if (curCtuid == null) {
+                    // we have to re-create and re-register the channel uuid
+                    logger.debug("Re-Registering channel type UUID for: {} ", shortName);
+                    var ct = buildAndRegisterChannelType(shortName, type, cx2e);
+                    var channelBuilder = ChannelBuilder.create(channel);
+                    channelBuilder.withType(ct.getUID());
+                    channel = channelBuilder.build(); // update channel
+                } else {
+                    // check if channel uuid still exists and re-carete when needed
+                    ChannelType ct = channelTypeProvider.getChannelType(curCtuid, null);
+                    if (ct == null) {
+                        buildAndRegisterChannelType(shortName, type, cx2e);
+                    }
+                }
+            } else if (channel == null || !Objects.equals(ctuid, channel.getChannelTypeUID())) {
                 logger.debug("Creating / updating channel {} of type {} for '{}'", shortName, channelType, description);
                 this.configChanged = true;
                 ChannelUID channelUID = new ChannelUID(this.taCmiSchemaHandler.getThing().getUID(), shortName);
@@ -456,15 +477,6 @@ public class ApiPageParser extends AbstractSimpleMarkupHandler {
                     logger.warn("Error configurating channel for {}: channeltype cannot be determined!", shortName);
                 }
                 channel = channelBuilder.build(); // add configuration property...
-            } else if (ctuid == null && cx2e != null) {
-                // custom channel type - check if it already exists and recreate when needed...
-                ChannelTypeUID curCtuid = channel.getChannelTypeUID();
-                if (curCtuid != null) {
-                    ChannelType ct = channelTypeProvider.getChannelType(curCtuid, null);
-                    if (ct == null) {
-                        buildAndRegisterChannelType(shortName, type, cx2e);
-                    }
-                }
             }
             this.configChanged = true;
             e = new ApiPageEntry(type, channel, address, cx2e, state);

--- a/bundles/org.openhab.binding.tacmi/src/main/java/org/openhab/binding/tacmi/internal/schema/ChangerX2Entry.java
+++ b/bundles/org.openhab.binding.tacmi/src/main/java/org/openhab/binding/tacmi/internal/schema/ChangerX2Entry.java
@@ -33,6 +33,7 @@ public class ChangerX2Entry {
     enum OptionType {
         NUMBER,
         SELECT,
+        TIME,
     }
 
     /**

--- a/bundles/org.openhab.binding.tacmi/src/main/java/org/openhab/binding/tacmi/internal/schema/ChangerX2Parser.java
+++ b/bundles/org.openhab.binding.tacmi/src/main/java/org/openhab/binding/tacmi/internal/schema/ChangerX2Parser.java
@@ -111,10 +111,49 @@ public class ChangerX2Parser extends AbstractSimpleMarkupHandler {
                 String type = attributes.get("type");
                 if ("number".equals(type)) {
                     this.optionType = OptionType.NUMBER;
-                    // we transfer the limits from the input elemnt...
+                    // we transfer the limits from the input element...
                     this.options.put(ChangerX2Entry.NUMBER_MIN, attributes.get(ChangerX2Entry.NUMBER_MIN));
                     this.options.put(ChangerX2Entry.NUMBER_MAX, attributes.get(ChangerX2Entry.NUMBER_MAX));
                     this.options.put(ChangerX2Entry.NUMBER_STEP, attributes.get(ChangerX2Entry.NUMBER_STEP));
+                } else {
+                    logger.warn("Error parsing options for {}: Unhandled input field in {}:{}: {}", channelName, line,
+                            col, attributes);
+                }
+            }
+        } else if ((this.parserState == ParserState.INIT || this.parserState == ParserState.INPUT)
+                && "input".equals(elementName) && "changetotimeh".equals(id)) {
+            this.parserState = ParserState.INPUT_DATA;
+            if (attributes != null) {
+                this.optionFieldName = attributes.get("name");
+                String type = attributes.get("type");
+                if ("number".equals(attributes.get("type"))) {
+                    this.optionType = OptionType.TIME;
+                    // validate hour limits
+                    if (!"0".equals(attributes.get(ChangerX2Entry.NUMBER_MIN))
+                            || !"24".equals(attributes.get(ChangerX2Entry.NUMBER_MAX))) {
+                        logger.warn(
+                                "Error parsing options for {}: Unexpected MIN/MAX values for hour input field in {}:{}: {}",
+                                channelName, line, col, attributes);
+                    }
+                    ;
+                } else {
+                    logger.warn("Error parsing options for {}: Unhandled input field in {}:{}: {}", channelName, line,
+                            col, attributes);
+                }
+            }
+        } else if ((this.parserState == ParserState.INPUT_DATA || this.parserState == ParserState.INPUT)
+                && "input".equals(elementName) && "changetotimem".equals(id)) {
+            this.parserState = ParserState.INPUT_DATA;
+            if (attributes != null) {
+                if ("number".equals(attributes.get("type"))) {
+                    this.optionType = OptionType.TIME;
+                    if (!"0".equals(attributes.get(ChangerX2Entry.NUMBER_MIN))
+                            || !"59".equals(attributes.get(ChangerX2Entry.NUMBER_MAX))) {
+                        logger.warn(
+                                "Error parsing options for {}: Unexpected MIN/MAX values for minute input field in {}:{}: {}",
+                                channelName, line, col, attributes);
+                    }
+                    ;
                 } else {
                     logger.warn("Error parsing options for {}: Unhandled input field in {}:{}: {}", channelName, line,
                             col, attributes);
@@ -136,6 +175,8 @@ public class ChangerX2Parser extends AbstractSimpleMarkupHandler {
             throws ParseException {
         if (this.parserState == ParserState.INPUT && "input".equals(elementName)) {
             this.parserState = ParserState.INIT;
+        } else if (this.parserState == ParserState.INPUT_DATA && "input".equals(elementName)) {
+            this.parserState = ParserState.INPUT;
         } else if (this.parserState == ParserState.SELECT && "select".equals(elementName)) {
             this.parserState = ParserState.INIT;
         } else if (this.parserState == ParserState.SELECT_OPTION && "option".equals(elementName)) {
@@ -159,6 +200,8 @@ public class ChangerX2Parser extends AbstractSimpleMarkupHandler {
                             channelName, line, col, value, prev, id);
                 }
             }
+        } else if (this.parserState == ParserState.INPUT && "span".equals(elementName)) {
+            // span's are ignored...
         } else {
             logger.debug("Error parsing options for {}: Unexpected CloseElement in {}:{}: {}", channelName, line, col,
                     elementName);

--- a/bundles/org.openhab.binding.tacmi/src/main/java/org/openhab/binding/tacmi/internal/schema/TACmiSchemaHandler.java
+++ b/bundles/org.openhab.binding.tacmi/src/main/java/org/openhab/binding/tacmi/internal/schema/TACmiSchemaHandler.java
@@ -39,6 +39,7 @@ import org.eclipse.jetty.client.api.Request;
 import org.eclipse.jetty.http.HttpHeader;
 import org.eclipse.jetty.http.HttpMethod;
 import org.openhab.binding.tacmi.internal.TACmiChannelTypeProvider;
+import org.openhab.core.library.types.DateTimeType;
 import org.openhab.core.library.types.OnOffType;
 import org.openhab.core.library.types.StringType;
 import org.openhab.core.thing.Channel;
@@ -257,6 +258,10 @@ public class TACmiSchemaHandler extends BaseThingHandler {
                     String val;
                     if (command instanceof Number qt) {
                         val = String.format(Locale.US, "%.2f", qt.floatValue());
+                    } else if (command instanceof DateTimeType dtt) {
+                        // time is transferred as minutes since midnight...
+                        var zdt = dtt.getZonedDateTime();
+                        val = Integer.toString(zdt.getHour() * 60 + zdt.getMinute());
                     } else {
                         val = command.format("%.2f");
                     }

--- a/bundles/org.openhab.binding.tacmi/src/main/java/org/openhab/binding/tacmi/internal/schema/TACmiSchemaHandler.java
+++ b/bundles/org.openhab.binding.tacmi/src/main/java/org/openhab/binding/tacmi/internal/schema/TACmiSchemaHandler.java
@@ -17,6 +17,7 @@ import java.nio.charset.StandardCharsets;
 import java.util.Base64;
 import java.util.HashMap;
 import java.util.List;
+import java.util.Locale;
 import java.util.Map;
 import java.util.concurrent.ExecutionException;
 import java.util.concurrent.ScheduledFuture;
@@ -253,8 +254,14 @@ public class TACmiSchemaHandler extends BaseThingHandler {
             case NUMERIC_FORM:
                 ChangerX2Entry cx2en = e.changerX2Entry;
                 if (cx2en != null) {
-                    reqUpdate = prepareRequest(buildUri("INCLUDE/change.cgi?changeadrx2=" + cx2en.address
-                            + "&changetox2=" + command.format("%.2f")));
+                    String val;
+                    if (command instanceof Number qt) {
+                        val = String.format(Locale.US, "%.2f", qt.floatValue());
+                    } else {
+                        val = command.format("%.2f");
+                    }
+                    reqUpdate = prepareRequest(
+                            buildUri("INCLUDE/change.cgi?changeadrx2=" + cx2en.address + "&changetox2=" + val));
                     reqUpdate.header(HttpHeader.REFERER, this.serverBase + "schema.html"); // required...
                 } else {
                     logger.debug("Got command for uninitalized channel {}: {}", channelUID, command);

